base/linalg/bidiag.jl

```diff
@@ -41,26 +41,30 @@ julia> ev = [7, 8, 9]
  9
 
 julia> Bu = Bidiagonal(dv, ev, :U) # ev is on the first superdiagonal
-4×4 Bidiagonal{Int64}:
+4×4 Bidiagonal{Int64,Array{Int64,1}}:
  1  7  ⋅  ⋅
  ⋅  2  8  ⋅
  ⋅  ⋅  3  9
  ⋅  ⋅  ⋅  4
 
 julia> Bl = Bidiagonal(dv, ev, :L) # ev is on the first subdiagonal
-4×4 Bidiagonal{Int64}:
+4×4 Bidiagonal{Int64,Array{Int64,1}}:
  1  ⋅  ⋅  ⋅
  7  2  ⋅  ⋅
  ⋅  8  3  ⋅
  ⋅  ⋅  9  4
 ```
 """
-function Bidiagonal(dv::AbstractVector{T}, ev::AbstractVector{T}, uplo::Symbol) where T
-    Bidiagonal{T}(collect(dv), collect(ev), char_uplo(uplo))
+function Bidiagonal(dv::V, ev::V, uplo::Symbol) where {T,V<:AbstractVector{T}}
+    Bidiagonal{T}(dv, ev, char_uplo(uplo))
+end
+
+function Bidiagonal(dv::AbstractVector{T}, ev::AbstractVector{S}, uplo::Symbol) where {T,S}
+    R = promote_type(T,S)
+    Bidiagonal(convert(AbstractVector{R}, dv), convert(AbstractVector{R}, ev), uplo)
 end
 
-function Bidiagonal(dv::AbstractVector{Td}, ev::AbstractVector{Te}, uplo::Symbol) where {Td,Te}
-    T = promote_type(Td,Te)
+function Bidiagonal(dv::AbstractVector{T}, ev::AbstractVector{T}, uplo::Symbol) where T
     Bidiagonal(convert(Vector{T}, dv), convert(Vector{T}, ev), uplo)
```

Comment:
This is just a fallback for cases like `Bidiagonal(::Vector{T}, ::GenericVector{T})`, but perhaps the constructor on line 62 should be generalized to promote to the same container type as well, and not just promote the element type? That would make this constructor obsolete.
Currently the promotion is as follows:
```julia
Bidiagonal(::Vector{T}, ::GenericVector{S}) ->
Bidiagonal(::Vector{R}, ::GenericVector{R}) ->
Bidiagonal(::Vector{R}, ::Vector{R})
```
The problem with generalizing the method above was this:
```julia
julia> A = rand(3); B = GenericArray(rand(3)); a, b = promote(A,B);

julia> typeof(a)
Array{Float64,1}

julia> typeof(b)
Base.Test.GenericArray{Float64,1}
```

i.e. they don't promote to the same container type. Hence the current scheme.